core/src/main/java/org/elasticsearch/index/query/QueryBuilder.java

```diff
@@ -72,4 +72,28 @@
      * Returns the name that identifies uniquely the query
      */
     String getName();
+
+    /**
+     * Rewrites this query builder into it's primitive form. By default this method return theb builder itself. If the builder
+     * did not change the identity reference must be returend otherwise the builder will be rewritten infinitely.
+     */
+    default QueryBuilder<?> rewrite(QueryRewriteContext queryShardContext) throws IOException {
+        return this;
+    }
+
+    /**
+     * Rewrites the given query into it's primitive form. Queries that for instance fetch resources from remote hosts or
+     * can simplify / optimize itself should do their heavy lifting duringt {@link #rewrite(QueryRewriteContext)}. This method
```

Comment:
s/it's primitive form/its primitive form    s/itself/themselves     s/duringt/during